File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        result=ListNode()
        r=result
        while (list1 and list2):
            r.next=ListNode()
            r=r.next
            if list1.val<list2.val:
                r.val=list1.val
                list1=list1.next
            else:
                r.val=list2.val
                list2=list2.next

        while list1:
            r.next=ListNode(val=list1.val)
            list1=list1.next
            r=r.next
        
        while list2:
            r.next=ListNode(val=list2.val)
            list2=list2.next
            r=r.next
        
        return result.next

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode* result=(struct ListNode*)malloc(sizeof(struct ListNode));
    result->next=NULL;
    struct ListNode* r=result;
    struct ListNode* p=list1;
    struct ListNode* q=list2;
    while(p && q){
        r->next=(struct ListNode*)malloc(sizeof(struct ListNode));
        r=r->next;
        r->next=NULL;
        if (p->val<q->val){
            r->val=p->val;
            p=p->next;
        }
        else{
            r->val=q->val;
            q=q->next;
        }
    }
    while(p){
        r->next=(struct ListNode*)malloc(sizeof(struct ListNode));
        r=r->next;
        r->next=NULL;
        r->val=p->val;
        p=p->next;
    }
    while(q){
        r->next=(struct ListNode*)malloc(sizeof(struct ListNode));
        r=r->next;
        r->next=NULL;
        r->val=q->val;
        q=q->next; 
    }
    return result->next;
}